backend/selectgen.ml

```diff
@@ -77,6 +86,21 @@ let env_enter_trywith env kind =
 let env_set_trap_stack env trap_stack =
   { env with trap_stack; }
 
+let cut_region_stack ~len rs =
+  let rs_len = List.length rs in
+  assert (len <= rs_len);
+  let rec cut trim acc xs =
+    if trim = 0
+    then List.rev acc, xs
+    else
+      match xs with
+      | [] -> assert false
+      | x :: xs -> cut (trim - 1) (x :: acc) xs
+  in
+  let pre, suff = cut (rs_len - len) [] rs in
+  assert (List.length suff = len);
+  pre, suff
```

Comment:
Hmm. If we make it a generic list function in Misc, I'd feel the need to make it a bit faster in case it gets called with long lists. Let's wait until there's a second use before generalising?

What's wrong with the assertion? Verifying postconditions is useful, saves having to read through the code when thinking about the return value.

Reply:
The assertion implies that the author thinks this code is potentially reachable other than some invariant or other they are maintaining. It makes the reader think that the function above might return a list of the wrong length if given the wrong input. For instance, it made me read very carefully through the function looking for how we might hit this place.

Another way to say that is that assertions are almost entirely used for pre-conditions, so using them for a post-condition can cause confusion.

Reply:
(The assertion point is nitpicking though -- if you disagree you can feel free to leave it in)

Reply:
There is already `Misc.Stdlib.List.split_at` I think, it is already used in a lot of places.

Reply:
Thanks! I'll use that.